File: ein.py/src/ein/inference/monotonic/_serialise.py
"""Serialisation helpers + the timeline-jsonl mixin (split out of state_dump.py).

Leaf module: the engine-agnostic Fact / Firing / KnowledgeBase → ein-source /
JSON renderers, plus `_TimelineMixin` (the shared `00_timeline.jsonl` writer),
used by both `MonotonicDumper` (state_dump.py) and `LatticeDumper`
(_lattice_dump.py).
"""
from __future__ import annotations

import json
import time
from collections.abc import Callable
from dataclasses import fields
from typing import Any

from ein.ir.types import Atom, Int, Keyword, KwPair, SForm, String
from ein.kb.entities import Fact, Layer
from ein.kb.store import KnowledgeBase

# ── Serialisation helpers ────────────────────────────────────
#
# Migrated 2026-05-29 out of ``inference.tree.state_dump``. The
# renderers + JSON serialisers are engine-agnostic: they project a
# :class:`Fact` / :class:`Firing` / :class:`KnowledgeBase` into ein
# source text or machine-parseable JSON, used by both
# :class:`MonotonicDumper` and :class:`LatticeDumper` below.


def _arg_to_node(arg: Any) -> Any:
    """Lower a Fact arg to an IR node."""
    if isinstance(arg, Fact):
        return _fact_to_sform(arg, with_kwargs=False)
    if isinstance(arg, int):
        return Int(value=arg)
    return Atom(name=str(arg))


def _fact_to_sform(fact: Fact, *, with_kwargs: bool = True) -> SForm:
    """Render a Fact as ``(rel arg0 arg1 ... :source "..." :rule "..." :layer ...)``.

    Nested-Fact args are recursively lowered without their kwargs
    (so they read as bare ``(rel args...)`` inside the outer form).
    """
    args: list[Any] = [_arg_to_node(a) for a in fact.args]
    if with_kwargs:
        prov = fact.provenance
        if prov is not None:
            if prov.kind == "source" and prov.source:
                args.append(KwPair(
                    key=Keyword(name="source"),
                    value=String(value=prov.source),
                ))
            elif prov.kind == "rule" and prov.rule:
                args.append(KwPair(
                    key=Keyword(name="rule"),
                    value=String(value=prov.rule),
                ))
            elif prov.kind == "hypothesis":
                args.append(KwPair(
                    key=Keyword(name="hypothesis"),
                    value=Int(value=prov.branch or 0),
                ))
        # Layer kw-pair only when provenance-derivation wouldn't recover
        # it on reload (P1.7c S1.7c.1): a `:rule` fact derives REASONING
        # and a `:source` fact derives FACT for free, so `:layer` is
        # emitted only for the residue — a hypothesis/unannotated
        # REASONING fact, an unsourced FACT, a sourced ONTOLOGY fact.
        # Keeps the flat dump → reload layer round-trip exact (the
        # wrapped loader silently dropped this annotation).
        has_rule = prov is not None and prov.kind == "rule" and bool(prov.rule)
        has_source = (
            prov is not None and prov.kind == "source" and bool(prov.source)
        )
        derived = (
            Layer.REASONING if has_rule
            else Layer.FACT if has_source
            else Layer.ONTOLOGY
        )
        if fact.layer is not derived:
            args.append(KwPair(
                key=Keyword(name="layer"),
                value=Atom(name=fact.layer.value),
            ))
    return SForm(head=Atom(name=fact.relation_name), args=tuple(args))


def _kb_to_ein_text(kb: KnowledgeBase) -> str:
    """Render a KB as a **flat** sequence of ein forms (P1.7c S1.7c.3).

    The block wrappers are gone — each fact is a top-level form. Layer is
    carried per fact: ONTOLOGY facts (the ``(relation ...)``, ``(is-a ...)``,
    ``(bijective ...)`` schema) carry no annotation; FACT facts carry
    ``:source``; REASONING facts (everything the saturator derived,
    including ``(not ...)``) carry ``:rule`` / ``:using`` — or an explicit
    ``:layer`` when provenance alone wouldn't re-derive the layer (see
    :func:`_fact_to_sform`). ONTOLOGY facts are emitted first purely for
    readability; the round-trip is order-independent.
    """
    from ein.ir.dump import dump_canonical

    ont: list[SForm] = []
    rest: list[SForm] = []
    for f in kb.facts:
        (ont if f.layer is Layer.ONTOLOGY else rest).append(_fact_to_sform(f))
    return dump_canonical([*ont, *rest])


def _firing_to_dict(firing: Any) -> dict[str, Any]:
    """Serialise a Firing for JSONL output.

    Stripped-down: rule name, activator, derived fact id,
    redundancy flag, and the premise fact-ids. Bindings flattened
    to ``{str: str}`` (Fact bindings dropped — chase them via
    premises).
    """
    bindings_clean: dict[str, str] = {}
    for k, v in firing.bindings.items():
        if isinstance(v, Fact):
            bindings_clean[k] = f"<fact:{v.relation_name}>"
        else:
            bindings_clean[k] = str(v)
    return {
        "rule": firing.rule,
        "activator": list(firing.activator),
        "bindings": bindings_clean,
        "redundant": firing.redundant,
        # S1.8.A13: `firing.derived` is now a tuple; this debug dump shows the
        # primary conclusion (the full fan-out is in the slice diagram + the KB
        # dump). `a` is a nested Fact (octagon) iff it has `.relation_name`.
        "derived": {
            "relation": firing.derived[0].relation_name,
            "args": [
                {"relation": a.relation_name,
                 "args": list(map(str, a.args))}
                if hasattr(a, "relation_name") else str(a)
                for a in firing.derived[0].args
            ],
        },
        "premises": [
            {"relation": p.relation_name,
             "args": list(map(str, p.args))}
            for p in firing.premises
        ],
    }


def _fact_summary(fact: Fact) -> dict[str, Any]:
    """Recursive Fact → JSON-friendly dict.

    Nested-Fact args (e.g. inside ``(not (color-loc Green House-3))``)
    render as nested ``{"relation": ..., "args": [...]}`` dicts so the
    output is machine-parseable. Atoms / ints / strings stringify.
    """
    return {
        "relation": fact.relation_name,
        "args": [
            _fact_summary(a) if isinstance(a, Fact) else str(a)
            for a in fact.args
        ],
    }


class _TimelineMixin:
    """Shared ``00_timeline.jsonl`` + ``summary.json`` plumbing.

    :class:`MonotonicDumper` and :class:`LatticeDumper` both stream a
    JSONL timeline and write a final ``summary.json``; they carried
    verbatim copies of ``_emit_timeline`` / ``summary`` / ``close``
    (F-ENG-11). The only behavioural knob is the timeline record's
    ``json.dumps`` ``default=`` serialiser — Monotonic emits with the
    strict default (``None``), Lattice passes ``str`` because its
    records can carry non-JSON-native payloads (FactId tuples). That
    knob is :attr:`_timeline_json_default`.

    Hosts must provide the ``_timeline_fp`` / ``_timeline_seq`` /
    ``started_at`` / ``out_dir`` attributes (both dumpers do, as
    dataclass fields).
    """

    # default= for timeline records; None ⇒ json.dumps' strict default.
    _timeline_json_default: Callable[[Any], Any] | None = None

    def close(self) -> None:
        """Close the timeline file without emitting ``summary.json``.

        Called by abort paths (``BudgetExceededError``) so the timeline
        is flushed + closed when no final summary will be written;
        normal exits close it via :meth:`summary`. Idempotent.
        """
        if self._timeline_fp is not None:
            self._timeline_fp.close()
            self._timeline_fp = None

    def summary(self, verdict: Any, stats: Any) -> None:
        verdict_kind = type(verdict).__name__ if verdict is not None else None
        elapsed = time.time() - self.started_at
        try:
            stats_dict = {
                f.name: getattr(stats, f.name)
                for f in fields(stats)
            }
        except TypeError:
            stats_dict = {"repr": repr(stats)}

        if self.out_dir is not None:
            (self.out_dir / "summary.json").write_text(
                json.dumps({
                    "verdict": verdict_kind,
                    "elapsed_seconds": round(elapsed, 3),
                    "stats": stats_dict,
                }, indent=2, sort_keys=True, default=str),
            )
        self._emit_timeline(
            "summary",
            verdict=verdict_kind,
            elapsed_seconds=round(elapsed, 3),
        )
        if self._timeline_fp is not None:
            self._timeline_fp.close()
            self._timeline_fp = None

    def _emit_timeline(self, event: str, **fields_: Any) -> None:
        if self._timeline_fp is None:
            return
        rec = {
            "seq": self._timeline_seq,
            "ts_ms": round((time.time() - self.started_at) * 1000, 3),
            "event": event,
            **fields_,
        }
        self._timeline_fp.write(
            json.dumps(rec, default=self._timeline_json_default) + "\n",
        )
        self._timeline_fp.flush()
        self._timeline_seq += 1

File: ein.rs/crates/ein-render/src/dump/serialise.rs
//! The dumpers' shared serialisers — `ein.py`'s
//! `inference/monotonic/_serialise.py`.
//!
//! Engine-agnostic projections of a fact, a firing or a whole KB into ein
//! source text or machine-parseable JSON, plus the `00_timeline.jsonl` writer
//! both file dumpers share.

use std::fs::File;
use std::io::Write;
use std::path::{Path, PathBuf};

use ein_core::{FactId, Kb, ProvKind, Tag, Terms, Value};
use ein_infer::firing::Firing;
use ein_ir::{Ast, Node, NodeId, dump_canonical};

use super::json::{Json, dumps, dumps_indent_sorted};

/// Lower a fact argument to an IR node.
fn arg_to_node(scratch: &mut Ast, terms: &Terms, arg: Value) -> NodeId {
    match arg.tag() {
        Tag::Fact => fact_to_sform(scratch, terms, None, arg.as_fact().expect("tagged Fact")),
        Tag::Int => {
            let sym = scratch.intern(terms.int_text(arg.as_int().expect("tagged Int")));
            scratch.push(Node::Int(sym), None)
        }
        // `Atom(name=str(arg))` — every non-fact, non-int argument.
        Tag::Sym => {
            let sym = scratch.intern(terms.sym(arg.as_sym().expect("tagged Sym")));
            scratch.push(Node::Atom(sym), None)
        }
    }
}

/// A fact as `(rel arg0 arg1 … :source "…" / :rule "…" / :hypothesis N)`.
///
/// A nested fact argument is lowered recursively **without** its keywords, so
/// it reads as a bare `(rel args…)` inside the outer form — which is why `kb`
/// is an `Option` rather than a plain reference: `None` means "no keywords".
fn fact_to_sform(scratch: &mut Ast, terms: &Terms, kb: Option<&Kb>, f: FactId) -> NodeId {
    let (rel, args) = terms.fact(f);
    let args = args.to_vec();
    let mut children: Vec<NodeId> = args
        .into_iter()
        .map(|a| arg_to_node(scratch, terms, a))
        .collect();
    if let Some(kb) = kb
        && let Some(prov) = kb.primary(f).map(|p| terms.provs.get(p))
    {
        let kw = |scratch: &mut Ast, key: &str, value: NodeId| {
            let name = scratch.intern(key);
            let key = scratch.push(Node::Keyword(name), None);
            scratch.push(Node::KwPair { key, value }, None)
        };
        match prov.kind {
            // Both guarded by truthiness in ein.py, so an empty string or a
            // missing name falls through to no keyword at all.
            ProvKind::Source => {
                if let Some(s) = prov.source.map(|s| terms.sym(s)).filter(|s| !s.is_empty()) {
                    let sym = scratch.intern(s);
                    let value = scratch.push(Node::Str(sym), None);
                    let pair = kw(scratch, "source", value);
                    children.push(pair);
                }
            }
            ProvKind::Rule => {
                if let Some(r) = prov.rule.map(|r| terms.sym(r)).filter(|r| !r.is_empty()) {
                    let sym = scratch.intern(r);
                    let value = scratch.push(Node::Str(sym), None);
                    let pair = kw(scratch, "rule", value);
                    children.push(pair);
                }
            }
            ProvKind::Hypothesis => {
                let sym = scratch.intern(&prov.branch.unwrap_or(0).to_string());
                let value = scratch.push(Node::Int(sym), None);
                let pair = kw(scratch, "hypothesis", value);
                children.push(pair);
            }
            ProvKind::Rejected => {}
        }
    }
    // The provenance *is* the origin, so the dump → reload round trip is
    // exact by construction rather than by patching.
    let name = scratch.intern(terms.sym(rel));
    let head = scratch.push(Node::Atom(name), None);
    scratch.sform(head, &children, None)
}

/// A KB as a **flat** sequence of ein forms.
///
/// The block wrappers are gone: each fact is a top-level form and where it
/// came from rides on its provenance. Un-annotated facts are emitted first
/// purely for readability; the round trip is order-independent.
pub fn kb_to_ein_text(kb: &Kb, terms: &Terms) -> String {
    let mut scratch = Ast::new();
    let (mut ont, mut rest): (Vec<NodeId>, Vec<NodeId>) = (Vec::new(), Vec::new());
    for f in kb.facts() {
        let node = fact_to_sform(&mut scratch, terms, Some(kb), f);
        // `f.is_given or f.is_derived` — given is a `:source`-kind with a
        // source, derived is any non-source kind.
        let prov = kb.primary(f).map(|p| terms.provs.get(p));
        let given = prov.is_some_and(|p| p.kind == ProvKind::Source && p.source.is_some());
        let derived = prov.is_some_and(|p| p.kind != ProvKind::Source);
        if given || derived {
            &mut rest
        } else {
            &mut ont
        }
        .push(node);
    }
    ont.extend(rest);
    dump_canonical(&scratch, &ont)
}

/// A firing for JSONL output — rule, activator, bindings, the primary derived
/// fact, and the premises.
pub fn firing_to_json(terms: &Terms, firing: &Firing) -> Json {
    let bindings: Vec<(String, Json)> = firing
        .bindings
        .iter()
        .map(|(k, v)| {
            let text = match v.tag() {
                // A `Fact` binding is dropped to a marker — chase it through
                // the premises instead.
                Tag::Fact => {
                    let (rel, _) = terms.fact(v.as_fact().expect("tagged Fact"));
                    format!("<fact:{}>", terms.sym(rel))
                }
                _ => terms.display(*v),
            };
            (terms.sym(*k).to_string(), Json::Str(text))
        })
        .collect();
    let fact_json = |f: FactId| {
        let (rel, args) = terms.fact(f);
        Json::obj(vec![
            ("relation", Json::str(terms.sym(rel))),
            (
                "args",
                Json::Array(args.iter().map(|a| Json::str(terms.display(*a))).collect()),
            ),
        ])
    };
    // The derived side shows the *primary* conclusion; a nested fact argument
    // renders as its own `{relation, args}` dict rather than as text.
    let (drel, dargs) = terms.fact(firing.derived[0]);
    let derived = Json::obj(vec![
        ("relation", Json::str(terms.sym(drel))),
        (
            "args",
            Json::Array(
                dargs
                    .iter()
                    .map(|a| match a.tag() {
                        Tag::Fact => fact_json(a.as_fact().expect("tagged Fact")),
                        _ => Json::str(terms.display(*a)),
                    })
                    .collect(),
            ),
        ),
    ]);
    Json::obj(vec![
        ("rule", Json::str(terms.sym(firing.rule))),
        (
            "activator",
            Json::Array(
                firing
                    .activator
                    .iter()
                    .map(|s| Json::str(terms.sym(*s)))
                    .collect(),
            ),
        ),
        ("bindings", Json::Object(bindings)),
        ("redundant", Json::Bool(firing.redundant)),
        ("derived", derived),
        (
            "premises",
            Json::Array(firing.premises.iter().map(|p| fact_json(*p)).collect()),
        ),
    ])
}

/// A fact as a recursive `{relation, args}` dict — nested facts nest, and
/// everything else stringifies.
pub fn fact_summary(terms: &Terms, f: FactId) -> Json {
    let (rel, args) = terms.fact(f);
    Json::obj(vec![
        ("relation", Json::str(terms.sym(rel))),
        (
            "args",
            Json::Array(
                args.iter()
                    .map(|a| match a.tag() {
                        Tag::Fact => fact_summary(terms, a.as_fact().expect("tagged Fact")),
                        _ => Json::str(terms.display(*a)),
                    })
                    .collect(),
            ),
        ),
    ])
}

/// The `00_timeline.jsonl` + `summary.json` plumbing both file dumpers share.
///
/// ein.py factors it as a mixin with one behavioural knob — the timeline
/// record's `json.dumps(default=)` serialiser. That knob does not survive the
/// port and does not need to: `default=` only fires for values Python's JSON
/// cannot encode, and every value built here is already encodable.
pub struct Timeline {
    pub out_dir: Option<PathBuf>,
    file: Option<File>,
    seq: u64,
    /// Seconds since the run started, for the timestamp fields — which are on
    /// the [normalisation list](../../../../plans/m1a_rust/design/01_parity_contract.md) §5,
    /// so what has to match is that they are *there* and well-shaped.
    started_at: std::time::Instant,
}

impl Timeline {
    /// Create the directory and open the timeline. `out_dir = None` skips
    /// every filesystem write while the hooks keep firing.
    pub fn new(out_dir: Option<&Path>) -> std::io::Result<Timeline> {
        let mut t = Timeline {
            out_dir: out_dir.map(Path::to_path_buf),
            file: None,
            seq: 0,
            started_at: std::time::Instant::now(),
        };
        if let Some(dir) = &t.out_dir {
            std::fs::create_dir_all(dir)?;
            std::fs::create_dir_all(dir.join("layers"))?;
            t.file = Some(File::create(dir.join("00_timeline.jsonl"))?);
        }
        Ok(t)
    }

    pub fn elapsed(&self) -> f64 {
        self.started_at.elapsed().as_secs_f64()
    }

    /// One timeline record: `seq`, `ts_ms`, `event`, then the caller's fields
    /// in the order given.
    pub fn emit(&mut self, event: &str, fields: Vec<(&str, Json)>) {
        let Some(file) = self.file.as_mut() else {
            return;
        };
        let mut rec = vec![
            ("seq".to_string(), Json::Int(self.seq as i64)),
            (
                "ts_ms".to_string(),
                Json::Float(round3(self.started_at.elapsed().as_secs_f64() * 1000.0)),
            ),
            ("event".to_string(), Json::str(event)),
        ];
        rec.extend(fields.into_iter().map(|(k, v)| (k.to_string(), v)));
        let _ = writeln!(file, "{}", dumps(&Json::Object(rec)));
        let _ = file.flush();
        self.seq += 1;
    }

    /// Write `summary.json` and close the timeline.
    pub fn summary(&mut self, verdict: &str, stats: Vec<(&str, Json)>) {
        let elapsed = round3(self.started_at.elapsed().as_secs_f64());
        if let Some(dir) = &self.out_dir {
            let doc = Json::obj(vec![
                ("verdict", Json::str(verdict)),
                ("elapsed_seconds", Json::Float(elapsed)),
                (
                    "stats",
                    Json::Object(stats.into_iter().map(|(k, v)| (k.to_string(), v)).collect()),
                ),
            ]);
            let _ = std::fs::write(dir.join("summary.json"), dumps_indent_sorted(&doc));
        }
        self.emit(
            "summary",
            vec![
                ("verdict", Json::str(verdict)),
                ("elapsed_seconds", Json::Float(elapsed)),
            ],
        );
        self.file = None;
    }

    /// Close the timeline without writing `summary.json` — the abort path, so
    /// the log is flushed when no final summary will land. Idempotent.
    pub fn close(&mut self) {
        self.file = None;
    }
}

/// `round(x, 3)` — Python's banker's rounding at three decimals.
fn round3(x: f64) -> f64 {
    let scaled = x * 1000.0;
    let r = scaled.round();
    // `f64::round` breaks ties away from zero; Python's `round` breaks them to
    // even. The difference only shows on an exact `.5`, which a clock reading
    // essentially never is — but "essentially never" is not "never".
    let r = if (scaled - scaled.trunc()).abs() == 0.5 && r % 2.0 != 0.0 {
        r - scaled.signum()
    } else {
        r
    };
    r / 1000.0
}
